okio/src/appleMain/kotlin/okio/BufferedSource.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+import kotlinx.cinterop.CPointer
+import kotlinx.cinterop.CPointerVar
+import kotlinx.cinterop.UnsafeNumber
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.convert
+import kotlinx.cinterop.pointed
+import kotlinx.cinterop.reinterpret
+import kotlinx.cinterop.usePinned
+import kotlinx.cinterop.value
+import platform.Foundation.NSData
+import platform.Foundation.NSError
+import platform.Foundation.NSInputStream
+import platform.Foundation.NSLocalizedDescriptionKey
+import platform.Foundation.NSUnderlyingErrorKey
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.darwin.NSUIntegerVar
+import platform.posix.memcpy
+import platform.posix.uint8_tVar
+
+fun BufferedSource.inputStream(): NSInputStream = BufferedSourceInputStream(this)
+
+/** Returns an input stream that reads from this source. */
+@OptIn(UnsafeNumber::class)
+private class BufferedSourceInputStream(
+  private val bufferedSource: BufferedSource
+) : NSInputStream(NSData()) {
+
+  private var error: NSError? = null
+
+  override fun streamError(): NSError? = error
+
+  override fun open() {
+    // no-op
+  }
+
+  override fun read(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+    try {
+      val internalBuffer = bufferedSource.buffer
+
+      if (bufferedSource is RealBufferedSource) {
+        if (bufferedSource.closed) throw IOException("closed")
+
+        if (internalBuffer.size == 0L) {
+          val count = bufferedSource.source.read(internalBuffer, Segment.SIZE.toLong())
+          if (count == -1L) return 0
+        }
+      }
+
+      val toRead = minOf(maxLength.toInt(), internalBuffer.size).toInt()
+      return internalBuffer.readNative(buffer, toRead).convert()
+    } catch (e: Exception) {
+      error = e.toNSError()
+      return -1
+    }
+  }
+
+  override fun getBuffer(
```

Comment:
I don’t think this is a big deal for now, but eventually it might be a problem ..  . .

Using usePinned to get the address of a byte within a buffer is fine, but I worry that the caller might need usePinned when they later read from the buffer.

Reply:
(This is only an issue if Kotlin’s GC ever learns to relocate buffers)

Reply:
I wondered if this could be problematic, but couldn't think of a way to pass the unpinning to the caller. I figure typical usage would read the buffer immediately after calling the function and wouldn't hold the reference longer. Still a minute chance GC occurs right between unpinning and reading though, and even tinier chance the address becomes invalid.

Copying the buffer would nullify the usefulness of the API and violate the documented behavior of [returning in 0(1)](https://developer.apple.com/documentation/foundation/nsinputstream). I figured having it implemented is probably preferable to not, but alternatively we could just return false.

I wonder what the behavior would be to `pin()` the address without `unpin()`ing it. Would GC still collect it once the reference is released, just disallow relocating it? Or would this create a memory leak?

```
val pinned = s.data.pin()
buffer?.pointed?.value = pinned.addressOf(s.pos).reinterpret()
length?.pointed?.value = (s.limit - s.pos).convert()
return true
```

Reply:
The memory leak question is also a good one. At the moment this API doesn't tell the GC about this reference to the ByteArray, so it could be collected (or recycled) before the caller gets to use it.

Maybe we should pin it here, and/or take other steps to avoid recycling, then unpin either on the next call to this method or on close(). That would work and it wouldn't leak.